src/api.ts

```diff
@@ -0,0 +1,387 @@
+  "FIREBASE_RULES_URL",
+  "https://firebaserules.googleapis.com"
+);
+export const runtimeconfigOrigin: Origin = utils.envOverride(
+  "FIREBASE_RUNTIMECONFIG_URL",
+  "https://runtimeconfig.googleapis.com"
+);
+export const storageOrigin: Origin = utils.envOverride(
+  "FIREBASE_STORAGE_URL",
+  "https://storage.googleapis.com"
+);
+export const firebaseStorageOrigin: Origin = utils.envOverride(
+  "FIREBASE_FIREBASESTORAGE_URL",
+  "https://firebasestorage.googleapis.com"
+);
+export const hostingApiOrigin: Origin = utils.envOverride(
+  "FIREBASE_HOSTING_API_URL",
+  "https://firebasehosting.googleapis.com"
+);
+export const cloudRunApiOrigin: Origin = utils.envOverride(
+  "CLOUD_RUN_API_URL",
+  "https://run.googleapis.com"
+);
+export const serviceUsageOrigin: Origin = utils.envOverride(
+  "FIREBASE_SERVICE_USAGE_URL",
+  "https://serviceusage.googleapis.com"
+);
+
+export function setRefreshToken(t = ""): void {
+  refreshToken = t;
+}
+
+export function setAccessToken(t = ""): void {
+  accessToken = t;
+}
+
+export function getScopes(): string[] {
+  return commandScopes;
+}
+
+export function setScopes(s: string[] = []): void {
+  commandScopes = _.uniq(
+    _.flatten(
+      [
+        scopes.EMAIL,
+        scopes.OPENID,
+        scopes.CLOUD_PROJECTS_READONLY,
+        scopes.FIREBASE_PLATFORM,
+      ].concat(s)
+    )
+  );
+  logger.debug("> command requires scopes:", JSON.stringify(commandScopes));
+}
+
+// eslint-disable-next-line camelcase
+export async function getAccessToken(): Promise<string> {
+  if (accessToken) return accessToken;
+  const r = await require("./auth").getAccessToken(refreshToken, commandScopes);
+  return r.access_token;
+}
+
+export async function addRequestHeaders(reqOptions: any): Promise<any> {
+  // Runtime fetch of Auth singleton to prevent circular module dependencies
+  _.set(reqOptions, ["headers", "User-Agent"], `FirebaseCLI/${CLI_VERSION}`);
+  _.set(reqOptions, ["headers", "X-Client-Version"], `FirebaseCLI/${CLI_VERSION}`);
+  const accessToken = await getAccessToken();
+  _.set(reqOptions, "headers.authorization", `Bearer ${accessToken}`);
+  return reqOptions;
+}
+
+export async function request(
+  method: Method | string,
+  resource: string,
+  options: FirebaseRequestOptions
+): Promise<any> {
+  options = _.extend(
+    {
+      data: {},
+      resolveOnHTTPError: false, // by default, status codes >= 400 leads to reject
+      json: true,
+    },
+    options
+  );
+
+  if (!options.origin) {
+    throw new FirebaseError("Cannot make request without an origin", { exit: 2 });
+  }
+
+  if (!VALID_METHODS.has(method)) {
+    method = Method.GET;
+  }
+
+  if (options.query) {
+    resource = appendQueryData(resource, options.query);
+  }
+
+  if (method === Method.GET) {
+    resource = appendQueryData(resource, options.data);
+  }
+
+  const reqOptions: requestModule.OptionsWithUrl = {
+    method: method,
+    url: options.origin + resource,
+  };
+
+  if (method !== "GET") {
```

Comment:
If we're going to use the enum in this file, let's use it everywhere.

```suggestion
  if (method !== Method.GET) {
```

Reply:
Done.